### PR TITLE
DeepFlavour Negative Tagger + DeepFlavour 10X training

### PR DESCRIPTION
**Pull request by emilbols**
This PR is concerned with the implementation of the DeepFlavour negative tagger, which is used for estimating the rate of light jets being tagged as b jets. The goal of the negative tagger is to have the same light jet distribution as the full tagger, but filter out as many b-jets as possible at high discriminator values. This is done by inverting the sign of all distributions that are symmetric under sign inversion for light jets, but not for b-jets.

A description of the negative taggers in general can be found in these presentation at the BTV meeting [1] and at the RECO meeting [2].

The DeepFlavour negative tagger distribution is shown in [3]. The negative tagger discriminator values are shown from -1 to 0, and the full tagger is shown from 0 to 1. Red is b-jets, blue and light blue are light jets, and green is c-jets.

Originally this PR was a part of PR https://github.com/cms-sw/cmssw/pull/23206 , but it was filtered out due to merge conflicts. The review comments by @slava77 left in that PR has been dealt with here.

PR https://github.com/cms-sw/cmssw/pull/23206 also introduced the 94X training which was made a while back. While that PR was ongoing we produced a 10X training, which performs better, so in this PR we also add a new DeepFlavour model trained on 10X MC samples. It is therefore required that the PR  https://github.com/cms-data/RecoBTag-Combined/pull/14  is merged first. The ROC curves can be seen in [4]. Here the dashed line is c vs b and the full line is l vs b.

[1] https://indico.cern.ch/event/713923/contributions/2938152/attachments/1620844/2578707/NegativeTagger.pdf

[2] https://indico.cern.ch/event/732060/contributions/3020340/attachments/1660464/2659922/NegativeTaggerPR.pdf

[3]
deepflavournegativetagger

[4]
rocs

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test with cms-sw/cmsdist#4084

**Comment by slava77** [comment]
@cmsbuild please test with cms-sw/cmsdist#4086

File: RecoBTag/TensorFlow/interface/SecondaryVertexConverter.h [hunk 1]
```diff
@@ -13,7 +13,8 @@ namespace btagbtvdeep {
   
   void svToFeatures( const reco::VertexCompositePtrCandidate & sv,
 		     const reco::Vertex & pv, const reco::Jet & jet,
-		     SecondaryVertexFeatures & sv_features) ;
+		     SecondaryVertexFeatures & sv_features,
+		     const bool flip_ = false) ;
```
Comment (slava77) [inline RecoBTag/TensorFlow/interface/SecondaryVertexConverter.h]: please do not use trailing underscore for local variables, this is usually used for member data

File: RecoBTag/TensorFlow/plugins/DeepFlavourTagInfoProducer.cc [hunk 2]
```diff
@@ -209,20 +218,16 @@ void DeepFlavourTagInfoProducer::produce(edm::Event& iEvent, const edm::EventSet
               { return btagbtvdeep::sv_vertex_comparator(sva, svb, pv); });
     // fill features from secondary vertices
     for (const auto & sv : svs_sorted) {
-      if (reco::deltaR(sv, jet) > jet_radius_) continue;
+      if (Geom::deltaR(sv.position() - pv.position(), flip_ ? -jet_dir : jet_dir) > jet_radius_) continue;
```
Comment (slava77) [inline RecoBTag/TensorFlow/plugins/DeepFlavourTagInfoProducer.cc]: please stay with reco::deltaR.
Also, it makes more sense to use deltaR2 and save some CPU cycles

File: RecoBTag/TensorFlow/python/pfNegativeDeepFlavourTagInfos_cfi.py [hunk 3]
```diff
@@ -3,5 +3,7 @@
 from RecoBTag.TensorFlow.pfDeepFlavourTagInfos_cfi import pfDeepFlavourTagInfos
 
 pfNegativeDeepFlavourTagInfos = pfDeepFlavourTagInfos.clone(
-        shallow_tag_infos = cms.InputTag('pfDeepCSVNegativeTagInfos')
-        )
+    shallow_tag_infos = cms.InputTag('pfDeepCSVNegativeTagInfos'),
+    secondary_vertices = cms.InputTag('inclusiveCandidateNegativeSecondaryVertices'),
+    flip = cms.bool(True)
```
Comment (slava77) [inline RecoBTag/TensorFlow/python/pfNegativeDeepFlavourTagInfos_cfi.py]: please drop type specifications for all parameters which already exist.
This is a safer way to protect from parameter name mistakes and will also help in possible parameter migrations.

File: RecoBTag/TensorFlow/src/SecondaryVertexConverter.cc [hunk 4]
```diff
@@ -12,11 +12,12 @@ namespace btagbtvdeep {
...
 		     const reco::Vertex & pv, const reco::Jet & jet,
-		     SecondaryVertexFeatures & sv_features) {
+		     SecondaryVertexFeatures & sv_features,
+		     const bool flip) {
     
+    math::XYZVector jet_dir = jet.momentum().Unit();
     sv_features.pt = sv.pt();
-    sv_features.deltaR = catch_infs_and_bound(
-					      std::fabs(reco::deltaR(sv,jet))-0.5,
+    sv_features.deltaR = catch_infs_and_bound(std::fabs(Geom::deltaR(sv.position() - pv.position(), flip ? -jet_dir : jet_dir))-0.5,
```
Comment (slava77) [inline RecoBTag/TensorFlow/src/SecondaryVertexConverter.cc]: reco::deltaR

File: RecoBTag/TensorFlow/src/SecondaryVertexConverter.cc [hunk 5]
```diff
@@ -31,7 +32,11 @@ namespace btagbtvdeep {
     sv_features.d3d = d3d_meas.value();
     sv_features.d3dsig = catch_infs_and_bound(d3d_meas.value()/d3d_meas.error(),
 					      0,-1,800);
-    sv_features.costhetasvpv = vertexDdotP(sv,pv);
+    float costhetasvpv = vertexDdotP(sv,pv);
+    if(flip){
+      costhetasvpv = -costhetasvpv;
+    }
```
Comment (slava77) [inline RecoBTag/TensorFlow/src/SecondaryVertexConverter.cc]: this can as well be ```sv_features.costhetasvpv = (flip ? -1.f : 1.f)* vertexDdotP(sv,pv);```

File: RecoBTag/TensorFlow/plugins/DeepFlavourTagInfoProducer.cc [hunk 6]
```diff
@@ -269,6 +274,9 @@ void DeepFlavourTagInfoProducer::produce(edm::Event& iEvent, const edm::EventSet
     features.n_pf_features.resize(n_sorted.size());
 
 
+    const edm::Provenance *prov = shallow_tag_infos.provenance();
+    const edm::ParameterSet& psetFromProvenance = edm::parameterSet(*prov);
+    double negative_cut = (( psetFromProvenance.getParameter<edm::ParameterSet>("computer") ).getParameter<edm::ParameterSet>("trackSelection")).getParameter<double>("sip3dSigMax");
```
[the remark this answers is not in the document]
Comment (emilbols) [reply RecoBTag/TensorFlow/plugins/DeepFlavourTagInfoProducer.cc]: done here https://github.com/cms-sw/cmssw/commit/b62675266e3129ca3f77a62f2d7b29a8546d6f84#diff-778e7825bed1fc0ad81083affff99fe3R279

File: RecoBTag/TensorFlow/python/pfNegativeDeepFlavourTagInfos_cfi.py (hunk 3 above)
Comment (emilbols) [reply RecoBTag/TensorFlow/python/pfNegativeDeepFlavourTagInfos_cfi.py]: done here https://github.com/cms-sw/cmssw/commit/b62675266e3129ca3f77a62f2d7b29a8546d6f84#diff-9e547e1a61fc38bcd4dc1951e6eb618fR6

File: RecoBTag/TensorFlow/src/SecondaryVertexConverter.cc (hunk 4 above)
Comment (emilbols) [reply RecoBTag/TensorFlow/src/SecondaryVertexConverter.cc]: done here https://github.com/cms-sw/cmssw/commit/b62675266e3129ca3f77a62f2d7b29a8546d6f84#diff-dc4d48592ab74339ae57ac34bb57bd4eR20

File: RecoBTag/TensorFlow/src/SecondaryVertexConverter.cc (hunk 5 above)
Comment (emilbols) [reply RecoBTag/TensorFlow/src/SecondaryVertexConverter.cc]: Done here https://github.com/cms-sw/cmssw/commit/b62675266e3129ca3f77a62f2d7b29a8546d6f84#diff-dc4d48592ab74339ae57ac34bb57bd4eR35

**Comment by perrotta** [comment]
@cmsbuild please test with cms-sw/cmsdist#4086

**Comment by slava77** [comment]
@ferencek @imarches @emilbols 
please clarify on the plans to have the MC validation plots for the deepFlavour (ROC curves and such).
It's quite unfortunate that the most powerful tagger has no standard DQM/validation plots.

**Comment by slava77** [comment]
Looking at JetHT data both 2016E (wf 135.7611) and 2017F (wf 136.8311) show rather similar changes in all 6 discriminants, except for perhaps probuds [all assignments are assuming I got the index to discriminant conversion right; most of them make sense].

probuds 2017
all_sign1024vsorig_runjetht2017freminiaod136p8311c_min2 max-2 patjets_slimmedjets__pat_obj___pairdiscrivector__24__second

probuds 2016
all_sign1024vsorig_runjetht2016ereminiaodwf136p7611c_min2 max-2 patjets_slimmedjets__pat_obj___pairdiscrivector__24__second

prob b 2017:
all_sign1024vsorig_runjetht2017freminiaod136p8311c_min2 max-2 patjets_slimmedjets__pat_obj___pairdiscrivector__19__second

prob g 2017
all_sign1024vsorig_runjetht2017freminiaod136p8311c_min2 max-2 patjets_slimmedjets__pat_obj___pairdiscrivector__21__second

Compare this to 2017 ttbar MC (wf 1325.51)
prob b
all_sign1024vsorig_ttbar13reminiaodwf1325p51c_min2 max-2 patjets_slimmedjets__pat_obj___pairdiscrivector__19__second

prob g
all_sign1024vsorig_ttbar13reminiaodwf1325p51c_min2 max-2 patjets_slimmedjets__pat_obj___pairdiscrivector__21__second

prob uds
all_sign1024vsorig_ttbar13reminiaodwf1325p51c_min2 max-2 patjets_slimmedjets__pat_obj___pairdiscrivector__24__second

Somewhat superficially, all these changes look like an improvement in performance in line with the ROC curves in [4] in the PR description.

Memory and CPU use didn't change significantly based on wf 136.8311.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/23467	b626752
- code changes are in line with the PR description and the related review
- jenkins tests pass and comparisons with the baseline show differences only in the miniAOD outputs in the deepFlavour discriminants saved in the slimmedJets
- local tests using larger samples show roughly expected behavior https://github.com/cms-sw/cmssw/pull/23467#issuecomment-395887045.

This PR should be integrated together or after cms-sw/cmsdist#4086

**Comment by imarches** [comment]
@slava77 next task for Emil is to move DeepFlavour to Reco, as he discussed during the Reco meeting (unfortunately I could not connect, but I was told there was nobody against this plan). This will allow to have DeepFlavour more easily integrated in the standard validation workflow.

**Comment by slava77** [comment]
On 6/8/18 4:14 PM, imarches wrote:
> @slava77 <https://github.com/slava77> next task for Emil is to move
> DeepFlavour to Reco, as he discussed during the Reco meeting
> (unfortunately I could not connect, but I was told there was nobody
> against this plan). This will allow to have DeepFlavour more easily
> integrated in the standard validation workflow.
> 

Are there any present issues to include relevant parts of BTag
validation in the @miniAODValidation sequence?
IIUC, all necessary products are available at run time and the standard
plots can be made

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Comment by emilbols** [comment]
@slava77 @jmduarte I guess the strategy for backport of this to 94X, will be to wait for the 94X backport of the deepDoubleB Tagger first?

**Comment by slava77** [comment]
On 6/11/18 6:40 AM, emilbols wrote:
> @slava77 <https://github.com/slava77> @jmduarte
> <https://github.com/jmduarte> I guess the strategy for backport of this
> to 94X, will be to wait for the 94X backport of the deepDoubleB Tagger
> first?
> 

yes.
In case it is substantially simpler to combine in one PR, it can be
considered as well.

**Comment by swertz** [comment]
> Are there any present issues to include relevant parts of BTag validation in the @miniAODValidation sequence? IIUC, all necessary products are available at run time and the standard plots can be made

There has been no "standard" miniAOD BTag validation set up since the discriminators in miniAOD are directly copied from RECO and not recomputed. What we've been doing is from time to time manually run a sequence to re-compute the taggers using miniAOD products and check the performance agreement with RECO (not jet-by-jet). We can and will naturally include DeepFlavour into this workflow.

Now, if DeepFlavour is only available in miniAOD it would indeed be good to validate it there too (and we'll add it to the standard validation when it's included in RECO).
About the "miniAODValidation sequence", could you point me to where it's configured? To be clear, we're talking about a sample-vs-sample comparison within miniAOD, not an object-vs-object comparison between miniAOD and REOC?
Also, when is this sequence run? An issue is that the validation uses products that are not available in miniAOD, but maybe these are available when that sequence is run...

**Comment by slava77** [comment]
On 6/12/18 2:51 AM, Sébastien Wertz wrote:
>     Are there any present issues to include relevant parts of BTag
>     validation in the @miniaodvalidation sequence? IIUC, all necessary
>     products are available at run time and the standard plots can be made
> 
> There has been no "standard" miniAOD BTag validation set up since the
> discriminators in miniAOD are directly copied from RECO and not
> recomputed. What we've been doing is from time to time manually run a
> sequence to re-compute the taggers using miniAOD products and check the
> performance agreement with RECO (not jet-by-jet). We can and will
> naturally include DeepFlavour into this workflow.
> 
> Now, if DeepFlavour is only available in miniAOD it would indeed be good
> to validate it there too (and we'll add it to the standard validation
> when it's included in RECO).

Does this mean that the dependence on PUPPI will be eliminated?
IIRC, this is currently the main reason why DeepFlavour is not running
in RECO.
Also, there is a perhaps secondary issue related to using packed
candidates directly which leads to differences between what can be
running in RECO and what's in miniAOD.

> About the "miniAODValidation sequence", could you point me to where it's
> configured? To be clear, we're talking about a sample-vs-sample
> comparison within miniAOD, not an object-vs-object comparison between
> miniAOD and REOC?

yes


> Also, when is this sequence run? An issue is that the validation uses
> products that are not available in miniAOD, but maybe these are
> available when that sequence is run...
> 

You can check the available products based on the configuration of a
miniAOD-only job.
E.g. runTheMatrix.py -l 1325.51
By running it, you will also get an example of how the miniAOD DQM and
Validation sequences are setup.

**Comment by imarches** [comment]
Hi @slava77 yes, you remember correctly. The next PR we will perform, that will move DeepFlavour to the RECO is indeed removing also the Puppi dependence

**Comment by ferencek** [comment]
I'm also not very familiar with the miniAOD DQM and Validation sequence but since it is run in the same job as the miniAOD production, all the AOD products should be available which should make it possible to run the standard btag validation with a few tweaks here and there.

As @swertz pointed out, up to now all ak4 btag discriminators were produced in RECO and then stored in miniAOD so there was no strong reason to implement anything in the miniAOD validation sequence. DeepFlavour is now an exception but if it will soon be in RECO, we will be back to the usual situation.

One place where miniAOD validation would be useful is fat jets for which we don't run b tagging in the standard reco.

File: RecoBTag/TensorFlow/plugins/DeepFlavourTagInfoProducer.cc [hunk 7]
```diff
@@ -209,20 +218,16 @@ void DeepFlavourTagInfoProducer::produce(edm::Event& iEvent, const edm::EventSet
               { return btagbtvdeep::sv_vertex_comparator(sva, svb, pv); });
     // fill features from secondary vertices
     for (const auto & sv : svs_sorted) {
-      if (reco::deltaR(sv, jet) > jet_radius_) continue;
+      if (reco::deltaR2(sv.position() - pv.position(), flip_ ? -jet_dir : jet_dir) > (jet_radius_*jet_radius_)) continue;
```
Comment (slava77) [inline RecoBTag/TensorFlow/plugins/DeepFlavourTagInfoProducer.cc]: @emilbols 
just to keep a record for future references: this change is undone in #23781 
Please confirm that this change here was made as a mistake and not as a result of an update to the definitions used for new training.

Comment (emilbols) [reply RecoBTag/TensorFlow/plugins/DeepFlavourTagInfoProducer.cc]: @slava77 it was just a mistake. In the training it was always done in the previous way.
